enhancements/automated-service-ca-rotation.md

```diff
@@ -0,0 +1,246 @@
+    ensure that trust chaining is possible.
+  - This provides a 6 month grace period for consumers of serving certs and ca
+    bundles to be refreshed.
+
+- Generate an intermediate CA certificate with the same public key as the new CA
+  but signed with the private key of the current CA.
+  - This intermediate certificate should be included with newly-generated serving
+    certificates, and serves to bridge trust between the CA bundle of an
+    unrefreshed client (validating trust with the old CA bundle) and a service
+    serving with a cert generated from the new CA.
+
+- Generate an intermediate CA certificate with the same public key as the current CA
+  but signed with the private key of the new CA.
+  - This intermediate certificate should be included in the the new CA
+    bundle along with the new CA's primary certificate, and serves to
+    bridge trust between an unrefreshed server (serving with a cert
+    generated from the previous CA) and a refreshed client (validating
+    trust with the new CA bundle).
+
+- Compare the `AuthorityKeyId` of a serving cert with the `SubjectKeyId` of the
+  current CA to determine whether the serving cert was issued by the current CA.
+  - If the serving cert is determined not to have been issued by the current CA,
+    it needs to be regenerated.
+  - Previously, `Subject.CommonName` of a serving cert was compared with
+    `Issuer.CommonName` of the CA. Since ensuring trust between old and new CAs
+    requires that `Subject.CommonName` be reused, this comparison will no longer
+    be sufficient to determine lineage.
+  - This requires that all new certificates - whether CA or entity - have
+    `AuthorityKeyId` and `SubjectKeyId` set.
+  - In addition to ensuring regeneration when `AuthorityKeyId` and `SubjectKeyId`
+    are set, this comparison schema will also ensure regeneration of serving
+    certs without `SubjectKeyId` set (due to being generated by a CA without
+    `AuthorityKeyId` set) since an empty `SubjectKeyId` will not match a
+    populated `AuthorityKeyId`.
+
+- Provide an indication that rotation has taken place via an event.
+
+- Allow rotation to be forced when
+  `ServiceCA.Spec.ForceRotation.Reason` is set to a non-empty value
+  that does not match the value for the
+  `service.beta.openshift.io/forced-rotation-reason` annotation on the
+  signing secret.
+  - Setting an annotation on the signing secret ensures that a forced
+    rotation is triggered at most once per reason.
+  - If a forced rotation and expiry-triggered rotation coincide, only
+    the forced rotation should be performed.
+
+- Allow the duration of the grace period for continuing to trust the
+  previous CA to be overridden when forcing rotation by setting
+  `ServiceCA.Spec.ForceRotation.GraceDuration`.
+  - If set, the expiry of the intermediate certs bridging trust
+    between the pre-rotation serving certs and CA bundle will be set
+    to the time of rotation plus the provided duration.
```

Comment:
This is going to extend the validity length of the new CA as well?

I think if someone wants to override with a custom grace period, you should have two values, one to be the validity length of the new CA, and another that is the percentage of that time to trigger a rotation. This way the grace period can be customized independently of the expiration date. (someone might want a normal service ca time and short grace period if they know to restart their services or whatever.). This might be kind of overkill.

At the least you would want to make it clear that setting the grace duration also makes the CA live longer.

Reply:
The `GraceDuration` field is only intended to affect the duration of continuing to trust the existing CA to ensure that clients of the operator have sufficient time to start using the new ca bundle and serving certs.  This would be accomplished by setting `NotAfter` for the template of both intermediate certs to `time.Now().Add(gracefulDuration)`. Since the field is not intended to have change the validity bounds of the new CA, it has no impact on when the next time-based rotation would be performed.

Reply:
Makes sense, I was misreading it as saying that it was adding time on to the new CA bundle.